compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2901,11 +2901,13 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
       case closure(_, _, _) =>
       case _ =>
         val recovered = typed(qual)(using ctx.fresh.setExploreTyperState())
-        report.errorOrMigrationWarning(OnlyFunctionsCanBeFollowedByUnderscore(recovered.tpe.widen), tree.srcPos, from = `3.0`)
+        val msg = OnlyFunctionsCanBeFollowedByUnderscore(recovered.tpe.widen, tree)
+        report.errorOrMigrationWarning(msg, tree.srcPos, from = `3.0`)
         if (migrateTo3) {
           // Under -rewrite, patch `x _` to `(() => x)`
-          patch(Span(tree.span.start), "(() => ")
-          patch(Span(qual.span.end, tree.span.end), ")")
+          msg.actions
+            .flatMap(_.patches)
+            .map(actionPatch => patch(actionPatch.srcPos.span, actionPatch.replacement))
           return typed(untpd.Function(Nil, qual), pt)
```

Comment:
- If there were multiple actions in the message we probably wouldn't want to apply all of them but only the first one, so I would do `actions.head`
- Since we already have two cases and will likely have more I suggest adding a method `Rewrite.applyAction` that takes an action as input and call `patch`.

Reply:
Sure, those both sound good. I've made both of these changes.